src/java.base/share/classes/jdk/internal/vm/CDSParallelPreProcessor.java

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright 2020 Google, Inc.  All Rights Reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  This particular file is
+ * subject to the "Classpath" exception as provided in the LICENSE file
+ * that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package jdk.internal.vm;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicInteger;
+
+/**
+ * Support loading and pre-processing classes in parallel at CDS dump time.
+ *
+ * CDS dumping involves a parallel phase and a non-parallel phase. The static
+ * CDSParallelPreProcessor.preLoadAndProcess() method is invoked by the VM to
+ * enter the parallel phase.
+ *
+ * During the parallel phase, the classlist is split into a number of sublists
+ * according to DumpWithParallelism value and processed parallelly in different
+ * tasks. Classes on the list are loaded but not explicitly initialized, and
+ * then pre-processed. Pre-processing classes during the parallel phase includes
+ * linking class, verifying class (when required), and rewriting bytecodes.
+ * After all parallel tasks complete, the control is transferred back to the VM,
+ * which then enters the non-parallel phase to further process the loaded data
+ * and create the archive file.
+ */
+public class CDSParallelPreProcessor {
```

Comment:
If any number of threads will use CDSParallelPreProcessor as entry point in the future, do we need a more general name, such as CDSPreProcessor?

Reply:
It's the entry point for the parallel phase. CDSParallelPreProcessor is appropriate. If we want to use it for the non-parallel phase, it's good to change the class name when new features are added. Non-parallel entry points could be encapsulated in a separate class. How about let's keep that open for now.

I'm less concerned with renaming this class, since I don't think that would cause any conflicts or require adjustments for applying the pre-init patches on top of that.

Reply:
Okay, then I'm fine with this.